.github/workflows/test.yml

```diff
@@ -39,25 +58,32 @@ jobs:
       - run: ./fmt.sh
         name: "HLint via ./fmt.sh"
 
+      - name: Set some window specific things
+        if: matrix.os == 'windows-latest'
+        run: |
+          echo "CABAL_STORE_DIR=$SYSTEMDRIVE\\SR" >> $GITHUB_ENV
+          echo "CABAL_PKGS_DIR=~\\AppData\\cabal\\packages" >> $GITHUB_ENV
+
+      - name: Set some linux/macOS specific things
+        if: matrix.os != 'windows-latest'
+        run: |
+          echo "CABAL_STORE_DIR=~/.cabal/store" >> $GITHUB_ENV
+          echo "CABAL_PKGS_DIR=~/.cabal/packages" >> $GITHUB_ENV
+
       - name: Cache Cabal
         uses: actions/cache@v2
         env:
           cache-name: cache-cabal
         with:
           path: |
-            ~/.cabal/packages
-            ~/.cabal/store
+            ${{ env.CABAL_PKGS_DIR }}
+            ${{ env.CABAL_STORE_DIR }}
           key: ${{ runner.os }}-${{ matrix.ghc }}-build-${{ hashFiles('cabal.project') }}
-          restore-keys: |
-            ${{ runner.os }}-${{ matrix.ghc }}-bench-$${ hashFiles('cabal.project') }}
-            ${{ runner.os }}-${{ matrix.ghc }}-build-
-            ${{ runner.os }}-${{ matrix.ghc }}
 
```

Comment:
It would be good to preserve the fallback restore keys so that the bench cache can be reused if the build cache is not available (which can happen when we go over the 5GB cache limit)

Reply:
ok, but we have to be careful with `restore-keys`, i detected weird behaviours in my tests: the build was using always the same cache even if i changed the `cabal.project`, then it built all deps (i guess the cached ones were not suitable) and finally it did not save the new cache. So the build was ignoring `cabal.project`  and rebuilding always.

I guess that behaviour comes from the partial keys so i will restore only the first one in the hope it will not interfere.

Reply:
@jneira It looks like the first key uses `bench` instead of `build`. Maybe that's why it used the same cache (the second key match) again and again.

Reply:
So I suggest the following `restore-keys`:
```yaml
          restore-keys: |
            ${{ runner.os }}-${{ matrix.ghc }}-build-$${ hashFiles('cabal.project') }}
            ${{ runner.os }}-${{ matrix.ghc }}-build-
            ${{ runner.os }}-${{ matrix.ghc }}-
```

Reply:
It was on purpose, to share the cache between both workflows. I think that behaviour comes from `${{ runner.os }}-${{ matrix.ghc }}-build-` and/or `${{ runner.os }}-${{ matrix.ghc }}-`.

But maybe if the previous bench (maybe from other pr) built deps with incompatible ghc options and the new test workflow needs to rebuild and after that it doesn't save the new cache, it will trigger the error.

@pepeiborra what about use the same key in bot workflows and take r

Reply:
> ok, but we have to be careful with `restore-keys`, i detected weird behaviours in my tests: the build was using always the same cache even if i changed the `cabal.project`, then it built all deps (i guess the cached ones were not suitable) and finally it did not save the new cache. So the build was ignoring `cabal.project` and rebuilding always.

What tests did you perform?

If there is a match, the saved cache is reused and no resaving happens. This is the expected behaviour: 
> Once you create a cache, you cannot change the contents of an existing cache but you can create a new cache with a new key.
https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows#using-contexts-to-create-cache-keys

I don't know why Cabal built all deps, next time try running it in verbose mode and inspect the logs.

> I guess that behaviour comes from the partial keys so i will restore only the first one in the hope it will not interfere.

The behaviour of cache keys is well documented - if there is more than one match, the most recent match will be used:

> If there are multiple partial matches for a restore key, the action returns the most recently created cache.

```
${{ runner.os }}-${{ matrix.ghc }}-bench-$${ hashFiles('cabal.project') }}
```
There was a typo in the expression above. It should have been:
```
${{ runner.os }}-${{ matrix.ghc }}-bench-${{ hashFiles('cabal.project') }}
```

> @pepeiborra what about use the same key in both workflows and take ride of restore-keys in both config files?

Bench jobs build more dependencies than build jobs. The cache key should uniquely identify the build artefacts, so I don't think it's a good idea to reuse the same key for both jobs: it will result in non deterministic build times.

Reply:
>What tests did you perform?

I am afraid that some runs have been removed

1. Observe this run: https://github.com/jneira/haskell-language-server/runs/1662701834
 * The commit corrected the typo in the `restore-keys` you mentioned
 * It has a cache hit with the key `Linux-8.10.3-build-972434a34a4f27d94bcde6e728eb4aee874c9d6722a3fbe1de0596eb28967094`
 * That key was the same before and after changing cabal.project (i dont have the runs)
 * That run did rebuild all deps and did not save the cache due to `Cache hit occurred on the primary key Linux-8.10.3-build-972434a34a4f27d94bcde6e728eb4aee874c9d6722a3fbe1de0596eb28967094, not saving cache.`
   * That alone should not be possible, a cache hit of the primary key + rebuild
 * The [next run](https://github.com/jneira/haskell-language-server/runs/1662701834), force pushing the same commit, did hit the same cache key and rebuilt all deps again
2. Another build: https://github.com/jneira/haskell-language-server/runs/1665564490 where i only removed the `restore-keys`, without changing cabal.project
  * The cache key changed: Cache not found for input keys: `Linux-8.10.3-build-21cd88e91a1f5546ffae7d4a25cc95f362b677ce03af1aa3a70d2dc45ca4ae09`
     * That should not be possible without changing cabal.project
   * In a second run https://github.com/jneira/haskell-language-server/runs/1667803429 force pushing the same commit the correct cache was hit and the deps had not rebuilt

> The behaviour of cache keys is well documented 

I have reread two times the docs while i was running the tests (and waiting for the build to finish, so plenty of time 😝) and the behavior i've observed does not correspond with. So either i am missing something in the tests or the cache implementation has a bug related with `restore-keys` and it is handle them like the primary one, for our use case (no globs, only one file to hash).

Reply:
But i am gonna try to reproduce in this pr, using the ci here, and taking detailed notes about the commits and the runs, without force pushing to keep them in the process. I am afraid that it will take some time.

Reply:
1 can be explained as a missing datapoint in the cache key. Something changed that invalidated the Cabal store, and the cache key did not capture it. Do we know if e.g. Cabal captures environment variables in its store keys? In case of doubt, I would add the `test.yml` hash to the cache key, and a `restore-key` without it. 

2 can be explained as follows: we can only store 5GB in cache artefacts, so it is possible that the cache for `Linux-8.10.3-build-21cd88e91a1f5546ffae7d4a25cc95f362b677ce03af1aa3a70d2dc45ca4ae09` had been evicted.

I don't think that there are any bugs, just not enough diagnostics. We could extend the `test.yml` script with a `cabal configure -v` script to find out why Cabal is rebuilding things, and GitHub could allow to download cache keys so that these problems could be debugged locally

Reply:
Ok,not sure if that can explain all the events (how can the cache key itself change removing the `restore-keys` and keeping `cabal.project` without changes)  but let's see if it can be even reproduced.